pep-0706.rst

```diff
@@ -0,0 +1,585 @@
+will change, after raising ``DeprecationWarning`` for 2 releases
+(shortest deprecation period allowed in Python's
+:pep:`backwards compatibility policy <387>`).
+
+Additionally, code that relies on ``TarInfo`` object identity may break,
+see :ref:`706-offset`.
+
+
+Backporting & Forward Compatibility
+===================================
+
+This feature may be backported to older versions of Python.
+
+In CPython, we don't add warnings to patch releases, so the default
+filter should be changed to ``'fully_trusted'`` in backports.
+
+Other than that, *all* of the changes to ``tarfile`` should be backported, so
+``hasattr(tarfile, 'data_filter')`` becomes a reliable check for all
+of the new functionality.
+
+Note that CPython's usual policy is to avoid adding new APIs in security
+backports.
+This feature does not make sense without a new API
+(``TarFile.extraction_filter`` and the ``filter`` argument),
+so we'll make an exception.
+(See `Discourse comment 23149/16 <https://discuss.python.org/t/23149/16>`__
+for details.)
+
+Here are examples of code that takes into account that ``tarfile`` may or may
+not not have the proposed feature.
+
+When copying these snippets, note that setting ``extraction_filter``
+will affect subsequent operations.
+
+* Fully trusted archive::
+
+    my_tarfile.extraction_filter = (lambda member, path: member)
+    my_tarfile.extractall()
+
+* Use the ``'data'`` filter if available, but revert to Python 3.11 behavior
+  (``'fully_trusted'``) if this feature is not available::
+
+    my_tarfile.extraction_filter = getattr(tarfile, 'data_filter',
+                                           (lambda member, path: member))
+    my_tarfile.extractall()
+
+  (This is an unsafe operation, so it should be spelled out explicitly,
+  ideally with a comment.)
+
+* Use the ``'data'`` filter; *fail* if it is not available::
+
+    my_tarfile.extractall(filter=tarfile.data_filter)
+
+  or::
+
+    my_tarfile.extraction_filter = tarfile.data_filter
+    my_tarfile.extractall()
+
+* Use the ``'data'`` filter; *warn* if it is not available::
+
+   if hasattr(tarfile, 'data_filter'):
+       my_tarfile.extractall(filter='data')
+   else:
+       # remove this when no longer needed
+       warn_the_user('Extracting may be unsafe; consider updating Python')
+       my_tarfile.extractall()
+
+
+Security Implications
+=====================
+
+This proposal improves security, at the expense of backwards compatibility.
+In particular, it will help users avoid `CVE-2007-4559`_.
+
+
+How to Teach This
+=================
+
+The API, usage notes and tips for further verification will be added to
+the documentation.
+These should be usable for users who are familiar wth archives in general, but
+not with the specifics of UNIX filesystems nor the related security issues.
+
+
+Reference Implementation
+========================
+
+Draft implementation: https://github.com/python/cpython/compare/main...encukou:cpython:tarfile-dir-traversal-sq
```

Comment:
```suggestion
`Draft implementation <https://github.com/python/cpython/compare/main...encukou:cpython:tarfile-dir-traversal-sq>`__
```

Avoid an ugly, space-consuming and non-a11y-friendly raw URL (or you could give it an appropriate descriptive link title)

Reply:
I prefer to put the full link here. An inline link suggests it goes to additional information or sources. This one is basically the whole section, and it's not part of a paragraph that would be hard to scan.
The a11y-friendliness angle might be a concern, but IMO, general a11y guidelines are for running text. When I imagine listening to the PEP, the long form doesn't seem bad.

Reply:
I guess I can make it a paragraph though!